x-pack/plugins/alerting/server/routes/snooze_rule.test.ts

```diff
@@ -21,17 +21,10 @@ beforeEach(() => {
   jest.resetAllMocks();
 });
 
-const SNOOZE_END_TIME = '2025-03-07T00:00:00.000Z';
+// These tests don't test for future snooze time validation, so this date doesn't need to be in the future
+const SNOOZE_END_TIME = '2021-03-07T00:00:00.000Z';
 
 describe('snoozeAlertRoute', () => {
-  beforeAll(() => {
-    jest.useFakeTimers('modern');
```

Comment:
So it turns out Jest fake timers doesn't actually work for mocking the output of `Date` or `moment` several imports deep. We just didn't realize this because the `snooze_rule` test isn't actually validating that the snooze date is in the future.

I was able to use `jest.spyOn` to mock `Date.now()` for the dropdown UI tests, but I haven't been able to find a reliable solution to mock `new Date()` or `moment()` with no arguments. We can work around this by using `new Date(Date.now()` or `moment(Date.now())` in any code where we want to test against the current date.

Reply:
I'd love to help figure out a way, but I don't know if I understand the problem. Can you elaborate more on what you're trying to do and what's not working?

Reply:
There's code in the `snooze` API that uses Moment to determine if the `snoozeEndTime` is in the future:

```js
moment().isBefore(snoozeEndTime);
```

And also in the new `RuleStatusDropdown` to determine how far in the future `snoozeEndTime` is:

```js
moment(item.snoozeEndTime).fromNow(true)
```

We thought that using `jest.useFakeTimers` would enable us to change the return result of `new Date()` and `Date.now()`. For example, in this test we called `jest.setSystemTime(new Date(2020, 3, 1))` to set the value of `now` to March 2020 (which, now that I think about it, oh god, why would anyone want to make anything permanently March 2020, oh no oh no oh no).

Turns out, this did not actually work as expected. It seems like Jest's fake timers system only applies to `Date()` called within the Jest tests themselves, and not to any modules that the tests import.

We didn't realize that this was broken because `snooze_rule.test.ts` isn't actually testing the part of the `snooze` API that tries to check if the snooze end time is in the future. But we thought that maybe it was.

It appears that `moment.fromNow()` uses `Date.now()` to determine the value of `now`, as opposed to `moment()` which uses `new Date()`. `jest.spyOn` allows us to mock the value of `global.Date.now` throughout all imports, but mocking `global.Date` requires you to mock the **entire Date object** and there's no easy way to just override the constructor.

So therefore, we can work around that limitation by doing something like:

```js
moment(Date.now()).isBefore(snoozeEndTime);
```

in any code that we need to test.